File: teamdraft_app/app/services/draft_service/line_up.rb

module DraftService
  class LineUp
    def initialize(players: [])
      @players = players
    end

    def call
      return { success: false, formation:   [ [], [], [] ] } if @players.empty?

      sorted_players = @players.sort_by(&:positioning)
      total = sorted_players.length

      formation = if total <= 4
        organize_small_team(sorted_players, total)
      else
        organize_full_team(sorted_players, total)
      end

      {
        success: true,
        formation: formation
      }
    end

    private

      def organize_small_team(sorted_players, total)
        mid_point = (total / 2.0).ceil
        [
          sorted_players[0...mid_point],    # Defense gets extra player if odd
          [],                               # No midfield
          sorted_players[mid_point..-1]     # Attack
        ]
      end

      def organize_full_team(sorted_players, total)
        base_size = total / 3
        remainder = total % 3

        def_size = base_size + (remainder > 0 ? 1 : 0)     # Defense gets first extra
        mid_size = base_size + (remainder > 1 ? 1 : 0)     # Midfield gets second extra

        current_index = 0
        [
          sorted_players[current_index...(current_index += def_size)],
          sorted_players[current_index...(current_index += mid_size)],
          sorted_players[current_index..-1]
        ]
      end
  end
end

File: js/lineup.js
/* TeamDraft v2 — lineup / formation
 * Splits a team into [defense, midfield, attack] rows by positioning,
 * mirroring the original DraftService::LineUp logic.
 */
window.TD = window.TD || {};

TD.lineup = (function () {
  const RANGES = { defensive: [0, 35], midfield: [36, 65], attacking: [66, 100] };

  function positionOf(p) {
    const v = Number(p.positioning);
    if (v <= 35) return "DEF";
    if (v <= 65) return "MID";
    return "ATT";
  }

  // Returns [defenseRow, midfieldRow, attackRow].
  function formation(players) {
    const sorted = players.slice().sort((a, b) => a.positioning - b.positioning);
    const total = sorted.length;
    if (total === 0) return [[], [], []];

    if (total <= 4) {
      const mid = Math.ceil(total / 2);
      return [sorted.slice(0, mid), [], sorted.slice(mid)];
    }
    const base = Math.floor(total / 3);
    const rem = total % 3;
    const defSize = base + (rem > 0 ? 1 : 0);
    const midSize = base + (rem > 1 ? 1 : 0);
    let i = 0;
    const def = sorted.slice(i, (i += defSize));
    const midR = sorted.slice(i, (i += midSize));
    const att = sorted.slice(i);
    return [def, midR, att];
  }

  return { formation, positionOf, RANGES };
})();
